File: algorithm/go/spiral-matrix-ii/p59.go
package main

import "fmt"

func generateMatrix(n int) [][]int {
	ret := make([][]int, n)
	for i := 0; i < n; i++ {
		ret[i] = make([]int, n)
	}
	if n == 0 {
		return ret
	}

	// l is square length
	// i is square No
	// val is value
	for l, i, val := n, 0, 1; val <= n*n; i++ {
		if l == 1 {
			ret[i][i] = val
			break
		}

		for k := 0; k < l-1; k++ {
			ret[i][i+k] = val
			val++
		}
		for k := 0; k < l-1; k++ {
			ret[i+k][i+l-1] = val
			val++
		}
		for k := 0; k < l-1; k++ {
			ret[i+l-1][i+l-1-k] = val
			val++
		}
		for k := 0; k < l-1; k++ {
			ret[i+l-1-k][i] = val
			val++
		}
		l -= 2
	}

	return ret
}
func main() {
	fmt.Println("vim-go")
}

File: algorithm/java/spiral-matrix-ii/P59.java
class Solution {
    public int[][] generateMatrix(int n) {
        int[][] ret = new int[n][n];
        if (n == 0) {
            return ret;
        }
        // n >= 2
        // loop by square, every square have 4 side;
        // if for n * n matrix
        // first square length is n
        // second square length is n-2
        // if n - 2 < 0 , there are not square;
        // first square from ret[0][0]
        // i-th square from ret[i-1][i-1]
        // let's simulate it
        for(int i = 0,len = n,val=1; len > 0; i++,len-=2) {
            // if len is 1, there are one point
            if (len == 1) {
                ret[i][i] = val;
                break;
            }
            // from (i, i) to  (i, i+len-2)
            for (int k =0; k <= len-2; k++) {
                ret[i][i+k] = val;
                val++;
            }            
            
            // from (i, i+len-1) to (i+len-2, i+len-1)
            for (int k=0; k <= len-2; k++) {
                ret[i+k][i+len-1]=val;
                val++;
            }
            
            // from (i+len-1, i+len-1)  to (i+len-1, i+1)
            for (int k=0; k <= len-2; k++) {
                ret[i+len-1][i+len-1-k] = val;
                val++;
            }
            // from (i+len-1, i) to (i+1, i)
            for (int k=0; k <= len-2; k++) {
                ret[i+len-1-k][i]=val;
                val++;
            }
        }
        
        
        return ret;
    }
}
